langchain4j-infinispan/src/main/java/dev/langchain4j/store/embedding/infinispan/InfinispanEmbeddingStore.java

```diff
@@ -0,0 +1,249 @@
+package dev.langchain4j.store.embedding.infinispan;
+
+import dev.langchain4j.data.document.Metadata;
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.store.embedding.EmbeddingMatch;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.configuration.ClientIntelligence;
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.commons.api.query.Query;
+import org.infinispan.commons.marshall.ProtoStreamMarshaller;
+import org.infinispan.protostream.FileDescriptorSource;
+import org.infinispan.protostream.SerializationContext;
+import org.infinispan.query.remote.client.ProtobufMetadataManagerConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.Utils.isNullOrEmpty;
+import static dev.langchain4j.internal.Utils.randomUUID;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotBlank;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+import static dev.langchain4j.internal.ValidationUtils.ensureTrue;
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Infinispan Embedding Store
+ */
+public class InfinispanEmbeddingStore implements EmbeddingStore<TextSegment> {
+
+    private static final Logger log = LoggerFactory.getLogger(InfinispanEmbeddingStore.class);
+
+    private final RemoteCache<String, LangchainInfinispanItem> remoteCache;
+
+    public RemoteCache<String, LangchainInfinispanItem> getRemoteCache() {
+        return remoteCache;
+    }
+    private final LanchainItemMarshaller itemMarshaller;
+
+    private static final String DEFAULT_CACHE_CONFIG =
+          "<distributed-cache name=\"CACHE_NAME\">\n"
+          +     "<indexing storage=\"local-heap\">\n"
+          +     "<indexed-entities>\n"
+          +     "<indexed-entity>LANCHAINITEM</indexed-entity>\n"
+          +     "</indexed-entities>\n"
+          +     "</indexing>\n"
+          + "</distributed-cache>";
+
+    private static final String PROTO = "syntax = \"proto2\";\n" + "\n" + "/**\n" + " * @Indexed\n" + " */\n"
+          + "message LanchainItemDIMENSION {\n" + "   \n" + "   /**\n" + "    * @Keyword\n" + "    */\n"
+          + "   optional string id = 1;\n" + "   \n" + "   /**\n" + "    * @Vector(dimension=DIMENSION, similarity=COSINE)\n"
+          + "    */\n" + "   repeated float floatVector = 2;\n" + "   \n" + "   optional string text = 3;\n" + "   \n"
+          + "   repeated string metadataKeys = 4;\n" + "   \n" + "   repeated string metadataValues = 5;\n" + "}\n";
```

Comment:
This is definitely not critical, but maybe it would be better to have the proto file more readable.

Reply:
the formatter did this, I'll leave it for now